### PR TITLE
soffice: Report keyboard-triggered paragraph style change

### PR DESCRIPTION
**Pull request by michaelweghorn**
### Link to issue number:

Partially implements feature requests from #6915

### Summary of the issue:

Commit 20bdb39e6f83ef87726817cb1bad31fa7b0f73cb
("soffice: Report keyboard-triggered font size change (#17147)") implemented announcement of the new value of the font size combobox in Writer's formatting toolbar when it is changed via keyboard shortcut. Issue #6915 requests announcement of further attributes, including the announcement of the current paragraph style when applying "Heading 1",..., "Heading 5" styles via the Ctrl+1,..., Ctrl+5 keyboard shortcuts.

### Description of user facing changes

When applying the "Body Text" or a heading paragraph style using the corresponding keyboard shortcut in LibreOffice Writer 25.2 or newer, NVDA announces the new paragraph style.

### Description of development approach

Extend the solution from the above-mentioned
commit 20bdb39e6f83ef87726817cb1bad31fa7b0f73cb.
Other than the simpler case of the font size combo box, changing the paragraph style can imply more related formatting changes (e.g. font size, bold,...).
To avoid triggering the announcement of these implicit formatting changes, but rather announce the new paragraph style to confirm that the user-triggered action had the expected result, introduce a way to reliably identify the paragraph style combobox in the formatting toolbar:

1) Add an "id" object attribute to the IAccessible2 object attribute specification:

> This identifier remains the same across different sessions of
> the same application, regardless of the UI language.

IAccessible2 commit: https://github.com/LinuxA11y/IAccessible2/commit/2b8c2c79417bad4b464761a142fab45ffde8bfa8

2) In LibreOffice, set the "id" attribute to the ID already used in the UI file format (GtkBuilder XML files): https://git.libreoffice.org/core/commit/21b29f25660db973fa1480de77e6a69d76a5de53

3) In NVDA, extend the logic for announcing gesture-triggered formatting changes to allow specifying a specific ID of the UI element(s) of interest. If one is set, don't announce changes in other UI elements for that gesture.
Set the ID to the one that the paragraph style combobox in Writer's formatting toolbar has ("applystyle").
Let NVDA handle the Ctrl+0, Ctrl+1, Ctrl+2, Ctrl+3, Ctrl+4 and Ctrl+5 gestures to announce the new paragraph style ("Body Text" for Ctrl+0, corresponding heading for Ctrl+1 through Ctrl+5).

### Testing strategy:

1. start NVDA
2. start current development version of LibreOffice Writer (with English UI and keyboard layout)
3. Press Ctrl+1 to apply paragraph Style "Heading 1"
4. Verify that NVDA announces the new paragraph style by name, i.e. says "Heading 1"
5. Repeat steps 3 and 4 with Ctrl+2, Ctrl+3, Ctrl+4, Ctrl+5 to verify that the other heading levels are also announced correctly.
6. Press Ctrl+0 to apply paragraph style "Body Text" and verify that NVDA announces the new paragraph style by name, i.e. says "Body Text"
8. Retest that the other supported keyboard gestures as implemented in commits 46a343685247ed7d1adfb737716590ec4be456a0 and 20bdb39e6f83ef87726817cb1bad31fa7b0f73cb are still announced as expected.

### Known issues with pull request:

None, but note that a current LibreOffice development version (to become LibreOffice 25.2) is required for the feature to actually work.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

@coderabbitai summary

### CODE REVIEW

File: source/appModules/soffice.py [hunk 1]
```diff
@@ -449,13 +457,36 @@ class SymphonyDocument(CompoundDocument):
 	# variables used for handling announcements resulting from gestures
 	GESTURE_ANNOUNCEMENT_TIMEOUT = 2.0
```
Comment (SaschaCowley) [inline source/appModules/soffice.py:458]: Could you please add the unit here?
```suggestion
	GESTURE_ANNOUNCEMENT_TIMEOUT = 2.0  # Seconds
```

File: source/appModules/soffice.py [hunk 2]
```diff
@@ -449,13 +457,36 @@ class SymphonyDocument(CompoundDocument):
...
 	announceFormattingGestureChange = False
+	formattingGestureObjectIds = []
 	lastFormattingGestureEventTime = 0
```
Comment (SaschaCowley) [inline source/appModules/soffice.py:461]: Can these please get type hints?

**Review by SaschaCowley — CHANGES_REQUESTED** [review]
Thanks for this, @michaelweghorn. Everything looks sound to me, I just have a request for a few type hints.

File: source/appModules/soffice.py [hunk 3]
```diff
@@ -46,6 +47,13 @@ def is_toolbar_item(obj: NVDAObject) -> bool:
 			parent = parent.parent
 		return False
 
+	@staticmethod
+	def get_id(obj: NVDAObject) -> Optional[str]:
```
Comment (seanbudd) [inline source/appModules/soffice.py:51]:
```suggestion
	def get_id(obj: NVDAObject) -> str | None:
```

File: source/appModules/soffice.py [hunk 4]
```diff
@@ -449,13 +457,36 @@ class SymphonyDocument(CompoundDocument):
 	# variables used for handling announcements resulting from gestures
 	GESTURE_ANNOUNCEMENT_TIMEOUT = 2.0
 	announceFormattingGestureChange = False
+	formattingGestureObjectIds = []
 	lastFormattingGestureEventTime = 0
 
 	@staticmethod
-	def isFormattingChangeAnnouncementEnabled() -> bool:
-		return SymphonyDocument.announceFormattingGestureChange and time.time() < (
+	def isFormattingChangeAnnouncementEnabled(obj: NVDAObject) -> bool:
```
Comment (seanbudd) [inline source/appModules/soffice.py:464]: this is an API breaking change, can it please be documented?

Comment (michaelweghorn) [reply source/appModules/soffice.py:464]: This method was only introduced recently in 20bdb39e6f83ef87726817cb1bad31fa7b0f73cb, i.e. if I understand correctly, it's only an API change within the current development version, not as compared to any released NVDA version. Should it still be documented?

File: source/appModules/soffice.py (hunk 2 above)
Comment (michaelweghorn) [reply source/appModules/soffice.py]: > Can these please get type hints?

Done

File: source/appModules/soffice.py (hunk 4 above)
Comment (seanbudd) [reply source/appModules/soffice.py:464]: Right, never mind

**Review by seanbudd — APPROVED** [review]

**Review by SaschaCowley — APPROVED** [review]

**Pull request merged**